csr/csr_to_socrata.py

```diff
@@ -0,0 +1,176 @@
+    df = pd.read_csv(ENDPOINT, sep="\t", encoding="utf_16")
+    logger.info(f"Downloaded {len(df)} CSRs from endpoint")
+    return df
+
+
+def convert_from_state_plane(df):
+    """
+    Adds WGS-84 lat/long columns to the dataframe based on the state plane coordinates.
+    """
+    gdf = gpd.GeoDataFrame(
+        df,
+        geometry=gpd.points_from_xy(
+            x=df["State Plane X Coordinate"], y=df["State Plane Y Coordinate"]
+        ),
+        crs="ESRI:102739",
+    )
+    gdf = gdf.to_crs("EPSG:4326")
+
+    # Get wgs84 location columns
+    df["latitude"] = gdf["geometry"].y
+    df["longitude"] = gdf["geometry"].x
+    df["location"] = df.apply(build_point_data, axis=1)
+    return df
+
+
+def build_point_data(row):
+    """
+
+    Parameters
+    ----------
+    row: dict of data from a pandas dataframe
+
+    Returns
+    -------
+    None if there is no location data given, or point datatype formatted as expected by Socrata.
+
+    """
+    if pd.isna(row["longitude"]) or pd.isna(row["latitude"]):
+        return None
+    return f"POINT ({row['longitude']} {row['latitude']})"
+
+
+def get_fiscal_year(row):
+    """
+    Returns the fiscal year based on the created date of the record
+    """
+    year = row["datetime"].year
+    month = row["datetime"].month
+    if month >= 10:
+        fiscal_year = year + 1
+    else:
+        fiscal_year = year
+
+    return fiscal_year
+
+
+def transform(df):
+    logger.info("Transforming CSR data")
+    df = convert_from_state_plane(df)
+
+    # Converting datetime to correct format for socrata
+    date_cols = [
+        "Status Change Date",
+        "Created Date",
+        "Overdue On Date",
+        "Last Update Date",
+        "Close Date",
+    ]
+    for col in date_cols:
+        df[col] = pd.to_datetime(df[col])
+        df[col] = df[col].dt.strftime("%Y-%m-%dT%H:%M:%S")
+        df[col] = df[col].where(pd.notnull(df[col]), None)
+
+    df["datetime"] = pd.to_datetime(df["Created Date"])
+    df["fiscal_year"] = df.apply(get_fiscal_year, axis=1)
+
+    # Field mapping
+    df = df[list(FIELD_MAPPING.keys())]
+    df.rename(columns=FIELD_MAPPING, inplace=True)
+
+    # Setting these NaN values to None, because Socrata will be mad otherwise
+    df["state_plane_x_coordinate"] = df["state_plane_x_coordinate"].replace(
+        {np.nan: None}
+    )
+    df["state_plane_y_coordinate"] = df["state_plane_y_coordinate"].replace(
+        {np.nan: None}
+    )
+    df["sr_location"] = df["sr_location"].replace({np.nan: None})
+
+    payload = df.to_dict("records")
+    return payload
+
+
+def load(client, data):
+    logger.info("Uploading CSR data to Socrata")
+    res = client.upsert(DATASET, data)
+    logger.info(res)
+    return res
+
+
+def main():
+    soda = Socrata(
+        SO_WEB,
+        SO_TOKEN,
+        username=SO_KEY,
+        password=SO_SECRET,
+        timeout=500,
+    )
+
+    data = extract()
+    data = transform(data)
+    res = load(soda, data)
```

Comment:
ETL - love it! i assume the pandas `pd.read_csv` method will `raise` if something goes wrong with that Socrata request? and sodapy also raises on HTTP exceptions if i recall correctly?

Reply:
@johnclary `pd.read_csv` will return a `UnicodeError` if it gets the wrong data type returned. This is because the web app firewall returns an HTML file. I'll add some additional logging for that one error. 

I believe those HTTP expections are caught with those sodapy requests.